src/Task/Rector.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+declare(strict_types=1);
+
+namespace GrumPHP\Task;
+
+use GrumPHP\Fixer\Provider\FixableProcessResultProvider;
+use GrumPHP\Runner\TaskResult;
+use GrumPHP\Runner\TaskResultInterface;
+use GrumPHP\Task\Context\ContextInterface;
+use GrumPHP\Task\Context\GitPreCommitContext;
+use GrumPHP\Task\Context\RunContext;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Process\Process;
+
+class Rector extends AbstractExternalTask
+{
+    public static function getConfigurableOptions(): OptionsResolver
+    {
+        $resolver = new OptionsResolver();
+        $resolver->setDefaults([
+            'config' => 'rector.php',
+            'triggered_by' => ['php'],
+            'ignore_patterns' => [],
+            'clear-cache' => true,
+            'no-progress-bar' => true,
+            'no-diffs' => false,
+        ]);
+
+        $resolver->addAllowedTypes('config', ['null', 'string']);
+        $resolver->addAllowedTypes('triggered_by', ['array']);
+        $resolver->addAllowedTypes('ignore_patterns', ['array']);
+        $resolver->addAllowedTypes('clear-cache', ['bool']);
+        $resolver->addAllowedTypes('no-progress-bar', ['bool']);
+        $resolver->addAllowedTypes('no-diffs', ['bool']);
+
+        return $resolver;
+    }
+
+    public function canRunInContext(ContextInterface $context): bool
+    {
+        return $context instanceof RunContext || $context instanceof GitPreCommitContext;
+    }
+
+    public function run(ContextInterface $context): TaskResultInterface
+    {
+        $config = $this->getConfig()->getOptions();
+
+        $files = $context
+            ->getFiles()
+            ->notPaths($config['ignore_patterns'])
+            ->extensions($config['triggered_by']);
+
+        if (0 === \count($files)) {
+            return TaskResult::createSkipped($this, $context);
+        }
+
+        $arguments = $this->processBuilder->createArgumentsForCommand('rector');
+        $arguments->add('process');
+        $arguments->add('--dry-run');
+        $arguments->add('--ansi');
+
+        $arguments->addOptionalArgument('--config=%s', $config['config']);
+        $arguments->addOptionalArgument('--clear-cache', $config['clear-cache']);
+        $arguments->addOptionalArgument('--no-progress-bar', $config['no-progress-bar']);
+        $arguments->addOptionalArgument('--no-diffs', $config['no-diffs']);
+
+        if ($context instanceof GitPreCommitContext) {
+            $arguments->addFiles($files);
+        }
+
+        $process = $this->processBuilder->buildProcess($arguments);
+        $process->run();
+
+        if (!$process->isSuccessful()) {
+            return FixableProcessResultProvider::provide(
+                TaskResult::createFailed($this, $context, $this->formatter->format($process)),
+                function () use ($arguments): Process {
+                    $arguments->add('--clear-cache');
```

Comment:
Any specific reason why you want to add --clear-cache possibly twice?
Wouldn't it make sense to use the cache here, since you just ran it without or decided to run without cache to start with?

Reply:
Indeed. Removed the option when fixing!